### PR TITLE
Low level fast parallel compressed checkpointing

### PR DESCRIPTION
**Pull request by kbowers-jump**
This implements low level APIs installed to facilitate high performance workspace checkpointing and fast snapshots.

- Currently needs proper updating of deps.sh to include the appropriate lz4 library (will work without it though ... compressed frames will not be supported).

- Also added a missing FD_HAS_ZSTD default to fd_util_base.h to match FD_HAS_LZ4 and fixed an #if / #ifdef confusion in fd_snapshot_istream.h.

### CODE REVIEW

**Review by ripatel-fd — APPROVED** [review]

**Review by ripatel-fd — DISMISSED** [review]

File: src/util/checkpt/fd_checkpt.c [hunk 1]
```diff
@@ -0,0 +1,911 @@
...
+
+           FIXME: DOES CHUNK_CSZ_MAX (DSTCAPACITY) AFFECT HOW LZ4 DOES
+           ITS COMPRESSION/DECOMPRESSION?  THE APIS SUGGEST NO AND THIS
+           ASSUMES IT DOESNT.  IF IT DOES, THIS CALL AND VARIOUS CALLS
+           BELOW MIGHT NEED TO BE ADJUST TO PROVIDE IDENTICAL VALUES FOR
+           DST_CAPACITY BETWEEN STREAMING AND MMIO AND BETWEEN CHECKPT
+           AND RESTORE. */
+
+        /* Note: 1 is default LZ4 acceleration */
+        int _chunk_csz = LZ4_compress_fast_continue( lz4, chunk, mmio + off, (int)chunk_usz, (int)chunk_csz_max, 1 );
```
Comment (ripatel-fd) [inline src/util/checkpt/fd_checkpt.c:369]: This is already known, but stating just for the record - The `LZ4_compress_fast_continue` is not safe with untrusted input

**Comment by lheeger-jump** [comment]
Will review shortly

File: src/util/checkpt/fd_checkpt.h [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+  uchar gbuf[ FD_RESTORE_PRIVATE_SBUF_SZ ]; /* gather optimization buffer */
+};
+
+/* fd_restore_t internals */
+
+struct fd_restore_private_rbuf { /* very similar to fd_io_buffered_istream */
+  uchar * mem;  /* Buffer of compressed bytes read from fd, byte indexed [0,rbuf_sz) */
+  ulong  sz;    /* Buffer size in bytes, >=FD_RESTORE_RBUF_MIN */
+  ulong  lo;    /* Buffer bytes [0,rbuf_lo) have been read and restored */
+  ulong  ready; /* Number of compressed bytes that haven't been processed, 0<=rbuf_lo<=(rbuf_lo+rbuf_ready)<=rbuf_sz */
```
Comment (ripatel-fd) [inline src/util/checkpt/fd_checkpt.h:191]: alignment

**Review by ripatel-fd — APPROVED** [review]

File: src/util/checkpt/fd_checkpt.h (hunk 2 above)
Comment (kbowers-jump) [reply src/util/checkpt/fd_checkpt.h:191]: None required (or really beneficial in this situation)

**Pull request merged**